### PR TITLE
[otlp] Switch to TagWriter for handling of tags/attributes

### PR DESCRIPTION
**Pull request by CodeBlanch**
Follow-up to #5476
Relates to #5475

## Changes

* Switch from `TagTransformer` to `TagWriter` design in OtlpExporter project

## Details

See #5475 for information on why this is being done. The TL;DR version is this is set up to write directly onto a stream/buffer and skip all allocations completely (future work). Also there is more utilization of the base class/less duplication of logic in the tag handling for array values inside OtlpExporter code.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated

### CODE REVIEW

File: src/Shared/TagWriter/TagWriter.cs [hunk 1]
```diff
@@ -80,7 +79,7 @@ internal abstract class TagWriter<TTagState, TArrayState>
             default:
                 try
                 {
-                    var stringValue = TruncateString(Convert.ToString(tag.Value, CultureInfo.InvariantCulture), tagValueMaxLength);
+                    var stringValue = TruncateString(Convert.ToString(tag.Value/*TODO: , CultureInfo.InvariantCulture*/), tagValueMaxLength);
```
Comment (vishweshbankwar) [inline src/Shared/TagWriter/TagWriter.cs:82]: Curious, why is this changed to TODO?

Comment (CodeBlanch) [reply src/Shared/TagWriter/TagWriter.cs:82]: The existing `TagTransformer<T>` we are migrating from doesn't use `InvariantCulture` today: https://github.com/open-telemetry/opentelemetry-dotnet/blob/eace49c2fb64656edf2de2ab9e8848596a3a3e40/src/Shared/TagTransformer.cs#L75

There is a unit test in OtlpExporter which is adding a `DateTime.UtcNow` as a tag value. With this change to use `InvariantCulture` that test fails because the format changes.

I discussed a bit with @alanwest. We feel this is a bug. We should be using `InvariantCulture` in the transformation code. But it is also a breaking change. So the idea was let's do this migration to `TagWriter<T>` without changing that behavior. Then we can do a dedicated PR/fix for that after the migration so it is independent/clearly documented.

**Review by vishweshbankwar — APPROVED** [review]

**Review by alanwest — APPROVED** [review]

**Review by rajkumar-rangaraj — APPROVED** [review]
